File: app/vision/thickness.py
"""Dickenkorrektur: Hoehenversatz zwischen Markerebene und Objektoberflaeche.

Eine Homographie ist nur fuer GENAU EINE Ebene exakt. Liegt das Markerblatt neben
dem Objekt auf dem Tisch, waehrend die interessante Flaeche h Millimeter hoeher
liegt, erscheint diese Flaeche radial vom Kamera-Lotpunkt weg gestreckt:

    T = N + k * (Q - N)        mit  k = d / (d - h)

T ist die scheinbare Lage in der Markerebene, Q die wahre Lage auf der
Objektoberflaeche, N der Lotpunkt, d die Kamerahoehe. Die Ruecktransformation ist
eine Skalierung um N mit 1/k; sie wird als Vorschaltmatrix in die Homographie
gefaltet, damit das Rasterbild direkt in wahren Objektkoordinaten entsteht.
"""

from __future__ import annotations

import numpy as np

from app.notices import AppError
from app.vision.camera import CameraPose


def correction_matrix(
    nadir_mm: tuple[float, float], camera_height_mm: float, thickness_mm: float
) -> np.ndarray:
    """Affine 3x3-Matrix, die wahre Objektkoordinaten auf ihre scheinbare Lage abbildet."""
    if abs(thickness_mm) < 1e-9:
        return np.eye(3)
    if camera_height_mm <= thickness_mm:
        raise AppError(
            "thickness_too_large",
            "thickness_mm",
            thickness_mm=f"{thickness_mm:.1f}",
            camera_height_mm=f"{camera_height_mm:.0f}",
        )

    factor = camera_height_mm / (camera_height_mm - thickness_mm)
    nadir_x, nadir_y = nadir_mm
    return np.array(
        [
            [factor, 0.0, nadir_x * (1.0 - factor)],
            [0.0, factor, nadir_y * (1.0 - factor)],
            [0.0, 0.0, 1.0],
        ]
    )


def effective_homography(
    homography: np.ndarray, pose: CameraPose, thickness_mm: float
) -> tuple[np.ndarray, float]:
    """Liefert (H_eff, k). Ohne Dicke ist H_eff = H und k = 1."""
    if abs(thickness_mm) < 1e-9:
        return np.asarray(homography, dtype=np.float64), 1.0

    if not pose.usable_for_thickness:
        raise AppError("camera_height_required", "camera_height_mm")

    assert pose.nadir_mm is not None and pose.height_mm is not None  # von usable_for_thickness
    correction = correction_matrix(pose.nadir_mm, pose.height_mm, thickness_mm)
    factor = pose.height_mm / (pose.height_mm - thickness_mm)
    return np.asarray(homography, dtype=np.float64) @ correction, float(factor)

File: web/vision/thickness.js
/**
 * thickness.js - Hoehenversatz zwischen Markerebene und Objektoberflaeche.
 * Gegenstueck zu app/vision/thickness.py.
 *
 * Eine Homographie ist nur fuer GENAU EINE Ebene exakt. Liegt das Markerblatt
 * neben dem Objekt auf dem Tisch, waehrend die interessante Flaeche h Millimeter
 * hoeher liegt, erscheint diese Flaeche radial vom Kamera-Lotpunkt weg gestreckt:
 *
 *     T = N + k * (Q - N)        mit  k = d / (d - h)
 *
 * T ist die scheinbare Lage in der Markerebene, Q die wahre Lage auf der
 * Objektoberflaeche, N der Lotpunkt, d die Kamerahoehe. Die Ruecktransformation
 * wird als Vorschaltmatrix in die Homographie gefaltet, damit das Rasterbild
 * direkt in wahren Objektkoordinaten entsteht.
 *
 * Reine Matrizenrechnung, deshalb bleibt sie hier und geht nicht in den Kern -
 * es gibt nichts, was zwischen zwei Umsetzungen driften koennte.
 */

import { matmul3 } from "./geometry.js";
import { AppError } from "./notices.js";
import { usableForThickness } from "./camera.js";

/** Affine 3x3-Matrix, die wahre Objektkoordinaten auf ihre scheinbare Lage abbildet. */
export function correctionMatrix(nadirMm, cameraHeightMm, thicknessMm) {
    if (Math.abs(thicknessMm) < 1e-9) {
        return new Float64Array([1, 0, 0, 0, 1, 0, 0, 0, 1]);
    }
    if (cameraHeightMm <= thicknessMm) {
        throw new AppError("thickness_too_large", "thickness_mm", {
            thickness_mm: thicknessMm.toFixed(1),
            camera_height_mm: cameraHeightMm.toFixed(0),
        });
    }

    const factor = cameraHeightMm / (cameraHeightMm - thicknessMm);
    const [nadirX, nadirY] = nadirMm;
    return new Float64Array([
        factor, 0.0, nadirX * (1.0 - factor),
        0.0, factor, nadirY * (1.0 - factor),
        0.0, 0.0, 1.0,
    ]);
}

/** Liefert { homography, factor }. Ohne Dicke ist H unveraendert und k = 1. */
export function effectiveHomography(homography, pose, thicknessMm) {
    if (Math.abs(thicknessMm) < 1e-9) {
        return { homography, factor: 1.0 };
    }
    if (!usableForThickness(pose)) {
        throw new AppError("camera_height_required", "camera_height_mm");
    }

    const correction = correctionMatrix(pose.nadirMm, pose.heightMm, thicknessMm);
    return {
        homography: matmul3(homography, correction),
        factor: pose.heightMm / (pose.heightMm - thicknessMm),
    };
}
